sycl/doc/extensions/fpga_io_pipes_design.rst

```diff
@@ -0,0 +1,88 @@
+.. code:: cpp
+  namespace intelfpga {
+    template <unsigned ID>
+    struct ethernet_pipe_id {
+      static constexpr unsigned id = ID;
+    };
+    using ethernet_read_pipe =
+        cl::sycl::intel::kernel_readable_io_pipe<ethernet_pipe_id<0>, int, 0>;
+    using ethernet_write_pipe =
+        cl::sycl::intel::kernel_writeable_io_pipe<ethernet_pipe_id<1>, int, 0>;
+  }
+
+  Thus, the user interacts only with vendor-defined pipe objects.
+
+Links
+-----
+.. _Spec: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/DataFlowPipes/data_flow_pipes.asciidoc
+.. _Interesting comment from Ronan: https://github.com/intel/llvm/pull/635#discussion_r325851766
+
+Summary
+-------
+This document describes a design of IO pipes implementation in SYCL compiler.
+It includes changes in SYCL library, clang and SPIR-V to LLVM IR translator.
+It adds extra attribute '__attribute__((io_pipe_id("access_mode", ID)))' that
+generates metadata, attached to a pipe storage declaration. By this metadata a
+backend can recognize IO pipe and distinguish different IO pipes from each
+other.
+
+There is another notable solution which was proposed by Ronan (see the link
+above): don't make any compiler/library changes - just make backends recognizing
+IO pipe by demangling its name. This proposal wasn't picked, because it will
+make backend support of the feature more difficult. So far we already have
+two devices' backends that support SYCL_INTEL_data_flow_pipes extension
+(Intel FPGA HW and Intel FPGA emulator) and in the future this number may
+increase. So efforts put in compiler implementation shall be payed off even
+more.
+
+clang
+-----
+Need to implement additional attribute, that possible to attach to a pipe
+storage declaration. The attribute shall accept a compiler time known integer
+argument (the pipe ID) and a string argument (access mode). With the attribute
+applied, clang generates a metadata attached to the pipe storage declaration,
+that contains the IO pipe ID and the access mode. The access modes shall be one
+of those: "read_only" and "write_only".
+
+llvm-spirv translator
+---------------------
+Need to implement additional decoration, that saves the IO pipe ID information
+and the access mode, that can be collected from the metadata attached to the
+pipe storage object.
+
+SYCL implementation in headers
+------------------------------
+Following the spec, we need to add two more classes for pipes:
+ - kernel_readable_io_pipe
+ - kernel_writeable_io_pipe
+
+ with the same member functions and fields as it is already done for pipe class.
+
+The attribute shall be attached to pipe storage declarations in the header, so
+the implementation shall be looking like:
+.. code:: cpp
+  static constexpr int32_t ID = name::id;
+  static constexpr struct ConstantPipeStorage
+  m_Storage __attribute__((io_pipe_id("read_only", ID))) =
+      {m_Size, m_Alignment, m_Capacity};
+
+
+, where 'name' is some class used in pipe type construction, like:
```

Comment:
Curious ` ,`

Reply:
Removed. I remember seeing such constructions in some science books written in Russian (which is for sure not making them grammar-wise correct neither in English nor in my native language), but for some reasons I'm still following this style.

Reply:
From a mathematical point of view, that makes sense! :-)